### ISSUE TITLE
How to modify HTTP response body?

### ISSUE THREAD

**Issue by soraliu** [solo-io/proxy-runtime#19]
I found that there are a lot of materials about modifying HTTP response headers. I want to inject some static codes into the HTTP response body, and how can I do it?

Hope somebody can help me.

**Comment by soraliu** [comment]
@yuval-k Hi man, may I ask for your help😂

**Comment by yuval-k** [comment]
Hi!
I'm not sure if it is possible with the current ABI; will have to check if the upstream ABI has updated to allow for this...

**Comment by RichiCoder1** [comment]
Would be very interested in this. Could potentially enable something like https://github.com/inikulin/parse5/blob/master/packages/parse5-html-rewriting-stream/docs/index.md or https://github.com/cloudflare/lol-html at the proxy level (assuming there's also a way to get body chunks)

**Comment by esnible** [comment]
My guess is that we are expected to return `StopIterationAndBuffer` from `onResponseBody()`, then later use the buffers in a call to `send_local_response()`.  I haven't gotten this working yet but that is the route I am attempting.

**Comment by RichiCoder1** [comment]
@esnible The appropriate APIs exists it looks like. See the Rust sdk for an example. https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/master/examples/http_body.rs

Specifically [`set_response_body`](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/master/src/traits.rs#L259) which appears to just be an abstraction of `set_buffer` which may or may not exist in this SDK?

**Comment by yuval-k** [comment]
i would not use send_local_reply for this purpose; i believe send_local_reply also sends the headers, so not sure if it can be called when orResponseBody happens;

what you probably want to do is manipulate the buffer as it goes through onResponseBody. this way you also minimize memory footprint

**Comment by esnible** [comment]
@yuval-k I have a user who wants to use WASM to recreate his current NGINX configuration which uses `error_page` and `proxy_pass` to fetch UI from one microservice to replace the error page generated by a different microservice.

I'm not an NGINX guy.  We thought maybe an EnvoyFilter could detect HTTP status codes on text/html pages, and if it saw a problem do httpCall() and use the results of that call as the body to the original call, keeping the original headers.

What do you think?

**Comment by ceastman-ibm** [comment]
Here is the current workaround that I have: @esnible and I are working on this together:

function envoy_on_response(response_handle)
  headers = response_handle:headers()
  streamInfo = response_handle:streamInfo()
  dynamicMetadata = streamInfo:dynamicMetadata()
  requestData = dynamicMetadata:get("request_headers")

-- This section of code is for redirecting to maintenance if a 4xx or 5xx status code is detected

  if string.match(headers:get(":status"), '[4-5].*') and requestData["accept"] ~= nil and string.find(string.lower(requestData["accept"]), "application/json") == nil then
    maintenance_url = "https://"..requestData["request_authority"].."/maintenance/"..headers:get(":status")
    headers:replace("location", maintenance_url)
    headers:replace(":status", "302")
  end

basically the goal is that the browser url doesn't change(redirect), it just displays a nice custom html page instead of the enovy test status page.

**Comment by cortex93** [comment]
Hi, any news on that ?

All other sdks (cpp, rust, go, zig) are importing `proxy_set_buffer_bytes ` to allow to modify the `BufferTypeValues.HttpResponseBody`.

The goal is to be able to do the same thing as in this envoy sample  https://github.com/envoyproxy/envoy/blob/6834081f835d609606e77ef84d734e30fc403d58/examples/wasm-cc/envoy_filter_http_wasm_updated_example.cc#L83

**Comment by yuval-k** [comment]
@cortex93 whatever you can do with other WASM envoy runtime, you should be able to do  with this one as well.

**Comment by cortex93** [comment]
> @cortex93 whatever you can do with other WASM envoy runtime, you should be able to do with this one as well.

I may have missed something but https://github.com/solo-io/proxy-runtime/blob/master/assembly/imports.ts is not importing proxy_set_buffer_bytes and this sdk is not exporting some helper as others sdk.

It seems clear to you but this thread is all about that and after more than a year, no one come with a working solution. Maybe you can share a sample ?

**Comment by kathuriasumit** [comment]
Hi @yuval-k, I am trying out the **proxy_set_buffer_bytes** implementation, This basically overrides the bytes in existing buffer, What if there is a need to enrich response body and required additional bytes in buffer? Do you have any suggestion for this use case?

**Comment by kathuriasumit** [comment]
> Hi @yuval-k, I am trying out the **proxy_set_buffer_bytes** implementation, This basically overrides the bytes in existing buffer, What if there is a need to enrich response body and required additional bytes in buffer? Do you have any suggestion for this use case?

Sharing below one of working approach. 
 ```ts
onResponseBody(body_buffer_length: usize, end_of_stream: bool): FilterDataStatusValues {
    if (!end_of_stream) {
      this.upstream_response_size = this.upstream_response_size + u32(body_buffer_length);
      return FilterDataStatusValues.StopIterationAndBuffer;
    }
    // read the response
    let original_response_body_buffer = get_buffer_bytes(BufferTypeValues.HttpResponseBody, 0, 
    this.upstream_response_size);
    let response_body_buffer = ...;
    // update the buffer
    set_buffer_bytes(BufferTypeValues.HttpResponseBody, 0, response_body_buffer.byteLength, response_body_buffer);
    return FilterDataStatusValues.Continue
    }
  
onResponseHeaders(a: u32, end_of_stream: bool): FilterHeadersStatusValues {
   // update the content length header
   stream_context.headers.response.replace("content-length", new_response_content_length.toString());
   }
    ```

**Comment by AleksFirsta** [comment]
Hi @kathuriasumit  Have you full example with updated repo? I want to realize ESI function with that
